### ISSUE TITLE
nixos/fail2ban: Make default settings optional/changable

### ISSUE THREAD

**Issue by OPNA2608** [NixOS/nixpkgs#62734]
## Issue description
I'd like to use settings for the DEFAULT and ssh-iptables jails that differ from those automatically set by the module, but there's currently no (easily discoverable?) way of doing so.
https://github.com/NixOS/nixpkgs/blob/92a047a6c4d46a222e9c323ea85882d0a7a13af8/nixos/modules/services/security/fail2ban.nix#L130-L148

When specifying different lines, the original ones get appended and override any duplicate setting.

I think wrapping these default jail definitions with `mkDefault` (which would have the drawback of throwing all default settings out the window if users want only 1 thing changed) or making more options for the most common settings + "extraConfig" would be better. :slightly_smiling_face: 

### Steps to reproduce
1)
```nix
services.fail2ban = {
  enable = true;
  jails.DEFAULT = ''
    bantime = 1d;
  '';
}
```
2) Check `/etc/fail2ban/jail.conf`.

There will be 2 `bantime` lines and the first one (yours) will get discarded during executing in favour of the second (default) one.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

 - system: `x86_64-linux`
 - host os: `Linux 5.1.5, NixOS, 19.03.172745.37698198717 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(bt1cn): `""`
 - channels(root): `"nixos-19.03.172745.37698198717"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by Ma27** [comment]
> I'd like to use settings for the DEFAULT and ssh-iptables jails that differ from those automatically set by the module, but there's currently no (easily discoverable?) way of doing so.

The common solution for overrides in the NixOS module system is to use [priorities](https://nixos.org/nixos/manual/#sec-option-definitions). In your case you could do something like this:

``` nix
{ lib, ... }:
{
  services.fail2ban = {
    enable = true;
    jails.DEFAULT = lib.mkForce ''
      # custom DEFAULT jail
    '';
  };
}
```

> I think wrapping these default jail definitions with mkDefault (which would have the drawback of throwing all default settings out the window if users want only 1 thing changed) or making more options for the most common settings + "extraConfig" would be better. slightly_smiling_face

`extraConfig` is IMHO not a good idea since you'll loose the entire merging capabilities you currently have which means that you can't remove or override jails defined by other modules.

I think we could introduce some kind of `recommendedJailDefaults` which defaults to true, similar too the ["recommended" options in the nginx module](https://nixos.org/nixos/options.html#services.nginx.recommended).

**Comment by OPNA2608** [comment]
> ```nix
> jails.DEFAULT = lib.mkForce ''
> ```

Ah, that sounds like it could be what I was looking for, thanks!

> `extraConfig` is IMHO not a good idea since you'll loose the entire merging  capabilities you currently have which means that you can't remove or  override jails defined by other modules.
>
>I think we could introduce some kind of `recommendedJailDefaults` which defaults to true, similar too the "recommended" options in the nginx module.

I think I was too vague, let me rephrase it.
What I meant was to keep the current setup with separate jails options with custom names, but to have `services.fail2ban.jails` function more like `users.users` with sub-options for the specific jail options (defaulting to the recommended settings for DEFAULT and to something like `null` for the other jails) instead of just one string for each jail.

I just don't know how many options there are, so I figured that, in addition to said option mapping, a `services.fail2ban.jails.<...>.extraConfig` to put any unmapped options in would be the easiest thing to do here. How much effort all of that would take to cleanly implement is something I haven't looked to deeply yet either.

Your `recommandedJailDefaults` option sounds acceptable as well though.

**Comment by Izorkin** [comment]
@OPNA2608 service fail2ban updated - #67931
Example worked variant:
```
   ignoreIP = [ "192.168.0.0/16" ];
    jails = {
      sshd = ''
        maxretry = 2
        mode     = aggressive
      '';
    };
```
